File: bal-app-code/clan-app/bal_app/lib/screens/focus/focus_cycle.dart
/// ═══════════════════════════════════════════════════════════
///  حساب طور دورة التركيز — نسخة التطبيق
///
///  ️ لازم تطابق `src/utils/focusCycle.js` في السيرفر **بالظبط**.
///     فيه اختبار (`test/focus-cycle-parity.test.mjs`) بيشغّل
///     النسختين على آلاف اللحظات ويقارن رقم برقم — لو حد عدّل
///     واحدة بس، الاختبار بيقع.
///
///  ليه محتاجين النسختين؟
///     السيرفر هو الحقيقة، بس منقدرش نسأله كل ثانية — ده هيولّع
///     البطارية ويغرق الشبكة. فبنسأله مرة واحدة عند الفتح ونعرف
///     `startedAt`، وبعدين نحسب محلياً من نفس النقطة. النتيجة:
///     العدّاد سلس، ولو المستخدم قفل التطبيق ساعة ورجع بيلاقي
///     نفسه في المكان الصح — مش واقف عند اللحظة اللي قفل فيها.
library;

enum FocusPhase { focus, rest, done }

class FocusPhaseState {
  final FocusPhase phase;
  final int cycleNumber;
  final Duration remaining;
  final Duration phaseTotal;
  final Duration totalRemaining;

  const FocusPhaseState({
    required this.phase,
    required this.cycleNumber,
    required this.remaining,
    required this.phaseTotal,
    required this.totalRemaining,
  });

  bool get isRest => phase == FocusPhase.rest;
  bool get isDone => phase == FocusPhase.done;

  /// نسبة اللي فاضل من الطور الحالي — للحلقة الدائرية
  double get progress {
    if (phaseTotal.inSeconds <= 0) return 0;
    return (remaining.inSeconds / phaseTotal.inSeconds).clamp(0.0, 1.0);
  }
}

abstract final class FocusCycle {
  /// الطول الكامل للجلسة — آخر دورة بلا راحة
  static Duration total({
    required int focusMin,
    required int restMin,
    required int cycles,
  }) {
    final f = (focusMin < 1 ? 1 : focusMin) * 60;
    final r = restMin < 0 ? 0 : restMin * 60;
    final n = cycles < 1 ? 1 : cycles;
    return Duration(seconds: n * (f + r) - r);
  }

  /// الطور عند لحظة معيّنة من بداية الجلسة.
  static FocusPhaseState at({
    required Duration elapsed,
    required int focusMin,
    required int restMin,
    required int cycles,
  }) {
    final f = (focusMin < 1 ? 1 : focusMin) * 60;
    final r = restMin < 0 ? 0 : restMin * 60;
    final n = cycles < 1 ? 1 : cycles;

    final cycleLen = f + r;
    final totalSec = n * cycleLen - r;
    final t = elapsed.inSeconds < 0 ? 0 : elapsed.inSeconds;

    //  خلصت — مش بنلفّ من الأول (ده كان باج في السيرفر)
    if (t >= totalSec) {
      return FocusPhaseState(
        phase: FocusPhase.done,
        cycleNumber: n,
        remaining: Duration.zero,
        phaseTotal: Duration(seconds: f),
        totalRemaining: Duration.zero,
      );
    }

    final cycleIdx = (t ~/ cycleLen) < n - 1 ? (t ~/ cycleLen) : n - 1;
    final inCycle = t - cycleIdx * cycleLen;
    final isLast = cycleIdx == n - 1;

    //  آخر دورة كلها تركيز
    if (isLast || inCycle < f) {
      return FocusPhaseState(
        phase: FocusPhase.focus,
        cycleNumber: cycleIdx + 1,
        remaining: Duration(seconds: f - inCycle),
        phaseTotal: Duration(seconds: f),
        totalRemaining: Duration(seconds: totalSec - t),
      );
    }

    return FocusPhaseState(
      phase: FocusPhase.rest,
      cycleNumber: cycleIdx + 1,
      remaining: Duration(seconds: cycleLen - inCycle),
      phaseTotal: Duration(seconds: r),
      totalRemaining: Duration(seconds: totalSec - t),
    );
  }
}

File: bal-app-code/clan-app/src/utils/focusCycle.js
/**
 * ═══════════════════════════════════════════════════════════
 *  حساب طور دورة التركيز — مصدر واحد للحقيقة
 *
 *  ️ ليه اتفصل في ملف لوحده:
 *
 *  الحساب ده كان موجود في `focus.controller.js` بس، والتطبيق كان
 *  بيعد **محلياً** بـ `Timer.periodic` من غير ما يسأل السيرفر
 *  خالص. النتيجة: لو المستخدم قفل التطبيق أو الشاشة نامت، العدّاد
 *  بيقف — بيرجع يلاقي نفسه في «دقيقة 3 من التركيز» والسيرفر شايفه
 *  في «الراحة التانية». رقمين مختلفين لنفس الجلسة.
 *
 *  دلوقتي الحساب هنا، والكنترولر بيستدعيه، والتطبيق بيعمل نفس
 *  الحساب من `startedAt` اللي السيرفر بعتهاله — فالاتنين بيوصلوا
 *  لنفس النتيجة مهما حصل. وفيه اختبار بيقارن النسختين رقم برقم.
 *
 *  شكل الدورة (مثال: 25 تركيز / 5 راحة / 3 دورات):
 *
 *    │ تركيز 25 │ راحة 5 │ تركيز 25 │ راحة 5 │ تركيز 25 │
 *    0         25       30        55       60        85
 *
 *  ️ آخر دورة **بلا راحة** — لأن الراحة بعد آخر تركيز مالهاش
 *    معنى، الجلسة خلصت. ده اللي بيخلي الطول = cycles*(f+r) - r
 * ═══════════════════════════════════════════════════════════
 */

/**
 * الطور الحالي في دورة مخصصة.
 *
 * @param {object} p
 * @param {number} p.elapsedSec      الثواني من بداية الجلسة
 * @param {number} p.focusMin        دقايق التركيز في الدورة الواحدة
 * @param {number} p.restMin         دقايق الراحة
 * @param {number} p.cycles          عدد الدورات
 * @returns {{
 *   name: 'FOCUS'|'REST'|'DONE',
 *   cycleNumber: number,
 *   remainingSec: number,
 *   phaseTotalSec: number,
 *   totalRemainingSec: number,
 * }}
 */
export const phaseAt = ({ elapsedSec, focusMin, restMin, cycles }) => {
  const f = Math.max(1, Math.round(focusMin)) * 60;
  const r = Math.max(0, Math.round(restMin)) * 60;
  const n = Math.max(1, Math.round(cycles));

  const cycleLen = f + r;
  const totalSec = n * cycleLen - r; //  آخر دورة بلا راحة
  const t = Math.max(0, Math.round(elapsedSec));

  //  الجلسة خلصت — مش بنلفّ من الأول
  if (t >= totalSec) {
    return {
      name: 'DONE',
      cycleNumber: n,
      remainingSec: 0,
      phaseTotalSec: f,
      totalRemainingSec: 0,
    };
  }

  const cycleIdx = Math.min(Math.floor(t / cycleLen), n - 1);
  const inCycle = t - cycleIdx * cycleLen;
  const isLast = cycleIdx === n - 1;

  //  آخر دورة كلها تركيز — مفيش راحة بعدها
  if (isLast || inCycle < f) {
    return {
      name: 'FOCUS',
      cycleNumber: cycleIdx + 1,
      remainingSec: f - inCycle,
      phaseTotalSec: f,
      totalRemainingSec: totalSec - t,
    };
  }

  return {
    name: 'REST',
    cycleNumber: cycleIdx + 1,
    remainingSec: cycleLen - inCycle,
    phaseTotalSec: r,
    totalRemainingSec: totalSec - t,
  };
};

/** الطول الكامل للجلسة بالثواني (بلا راحة أخيرة) */
export const totalSeconds = ({ focusMin, restMin, cycles }) => {
  const f = Math.max(1, Math.round(focusMin)) * 60;
  const r = Math.max(0, Math.round(restMin)) * 60;
  const n = Math.max(1, Math.round(cycles));
  return n * (f + r) - r;
};

export default { phaseAt, totalSeconds };
